File: src/ableton_mcp/protocol.py
"""Wire framing for the Live bridge: 4-byte big-endian length prefix + UTF-8 JSON.

This file must stay identical to remote_script/AbletonMCP/core/protocol.py
(the remote script has to be self-contained). A unit test enforces the sync.
"""

import json
import struct

HEADER = struct.Struct(">I")
MAX_FRAME_BYTES = 16 * 1024 * 1024


class FrameError(Exception):
    """Unrecoverable framing error. The connection should be closed."""


def encode(message):
    body = json.dumps(message, ensure_ascii=False).encode("utf-8")
    if len(body) > MAX_FRAME_BYTES:
        raise FrameError("frame of %d bytes exceeds limit" % len(body))
    return HEADER.pack(len(body)) + body


class FrameDecoder:
    """Accumulates bytes, yields decoded messages as complete frames arrive."""

    def __init__(self):
        self._buffer = bytearray()

    def feed(self, data):
        self._buffer.extend(data)
        messages = []
        while len(self._buffer) >= HEADER.size:
            (length,) = HEADER.unpack(bytes(self._buffer[: HEADER.size]))
            if length > MAX_FRAME_BYTES:
                raise FrameError("frame of %d bytes exceeds limit" % length)
            if len(self._buffer) < HEADER.size + length:
                break
            body = bytes(self._buffer[HEADER.size : HEADER.size + length])
            del self._buffer[: HEADER.size + length]
            try:
                messages.append(json.loads(body.decode("utf-8")))
            except (ValueError, UnicodeDecodeError) as exc:
                raise FrameError("invalid frame payload: %s" % exc)
        return messages

File: server/src/protocol.ts
// Wire framing for the Live bridge: 4-byte big-endian length prefix + UTF-8
// JSON. Must stay wire-compatible with remote_script/AbletonMCP/core/protocol.py.

const HEADER_SIZE = 4;
export const MAX_FRAME_BYTES = 16 * 1024 * 1024;

export class FrameError extends Error {}

export function encode(message: unknown): Buffer {
  const body = Buffer.from(JSON.stringify(message), "utf-8");
  if (body.length > MAX_FRAME_BYTES) {
    throw new FrameError(`frame of ${body.length} bytes exceeds limit`);
  }
  const frame = Buffer.alloc(HEADER_SIZE + body.length);
  frame.writeUInt32BE(body.length, 0);
  body.copy(frame, HEADER_SIZE);
  return frame;
}

export class FrameDecoder {
  private buffer: Buffer = Buffer.alloc(0);

  feed(data: Buffer): unknown[] {
    this.buffer = Buffer.concat([this.buffer, data]);
    const messages: unknown[] = [];
    while (this.buffer.length >= HEADER_SIZE) {
      const length = this.buffer.readUInt32BE(0);
      if (length > MAX_FRAME_BYTES) {
        throw new FrameError(`frame of ${length} bytes exceeds limit`);
      }
      if (this.buffer.length < HEADER_SIZE + length) {
        break;
      }
      const body = this.buffer.subarray(HEADER_SIZE, HEADER_SIZE + length);
      this.buffer = this.buffer.subarray(HEADER_SIZE + length);
      try {
        messages.push(JSON.parse(body.toString("utf-8")));
      } catch (error) {
        throw new FrameError(`invalid frame payload: ${error}`);
      }
    }
    return messages;
  }
}
